contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/recipes/ZKLock.java

```diff
@@ -0,0 +1,237 @@
+          } else {
+            // try acquire.
+            return acquire(deadline);
+          }
+        })
+        /* run the given task with toTry() */
+        .flatMap(unused -> task).toTry()
+        /* release the lock and unwind the result */
+        .flatMap(result -> release().andThen(unwind(result)));
+  }
+
+  /**
+   * Unwinds the result from {@link Task#toTry()}.
+   *
+   * @return task with the raw result.
+   */
+  private <T> Task<T> unwind(Try<T> result) {
+    if (result.isFailed()) {
+      return Task.failure(result.getError());
+    } else {
+      return Task.value(result.get());
+    }
+  }
+
+  /**
+   * Try to acquire the lock within the given deadline.
+   */
+  private Task<String> acquire(long deadline) {
+    final String uuid = UUID.randomUUID().toString();
+
+    return
+        /* Step 1: create znode with a pathname of "_locknode_/guid-lock-"
+         * and the sequence and ephemeral flags set. */
+        safeCreateLockNode(uuid, deadline, false)
+            .onFailure(e -> _zkClient.deleteNodeHasUUID(_lockPath, uuid))
+            .flatMap(lockNode -> tryAcquire(lockNode, deadline)
+                .withTimeout(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS)
+                .onFailure(e -> _zkClient.deleteNode(lockNode)));
+  }
+
+  /**
+   * Try to release the lock.
+   */
+  private Task<Void> release() {
+    return PlanLocal.get(getPlanLocalKey(), LockInternal.class).flatMap(lockInternal -> {
+      // should never be null.
+      if (lockInternal == null) {
+        LOG.error("LockInternal is null when releasing lock: ", _lockPath);
+      } else {
+        lockInternal._lockCount--;
+        if (lockInternal._lockCount == 0) {
+
+          return PlanLocal.remove(getPlanLocalKey())
+              .flatMap(unused ->
+                  _zkClient.delete(lockInternal._lockNode, -1)
+                      .recover(e -> {
+                        _zkClient.deleteNode(lockInternal._lockNode);
+                        return null;
+                      })
+              );
+        }
+      }
+      return Task.value(null);
+    });
+  }
+
+  private Task<String> tryAcquire(String lockNode, long deadline) {
+    return
+        /* Step 2: Call getChildren( ) on the lock node without setting the watch flag */
+        _zkClient.getChildren(_lockPath)
+            .flatMap(children -> {
+              children = children.stream().map(child -> ZKUtil.noramlizeZKPath(_lockPath, child))
+                  .collect(Collectors.toList());
+              String node = ZKUtil.findNodeWithNextLowestSN(children, lockNode);
+              if (node.equals(lockNode)) {
+                /* Step 3: If the pathname created in step 1 has the lowest sequence number suffix,
+                 * the client has the lock and the client exits the protocol. */
+                return PlanLocal.put(getPlanLocalKey(), new LockInternal(lockNode)).map(unused -> lockNode);
```

Comment:
the value part is weird...
it seems we just want to put lockNode, but why coupling with instantiation.